### ISSUE TITLE
Cryptic error "duplicate symbol: __foo.thunk.0 defined in <internal>"

### ISSUE THREAD

**Issue by nico** [llvm/llvm-project#54599]
1. Download https://drive.google.com/file/d/1C6gl6g6ovSTzfPVyEqgaFI4VM5ZaLz7Y/view?usp=sharing
2. `tar zxf repro_internal.tar.gz && cd repro`
3. ```
    ~/src/llvm-project/out/gn/bin/ld64.lld @response.txt
   ld64.lld: error: duplicate symbol: __ZN15partition_alloc8internal12_GLOBAL__N_126FreelistCorruptionDetectedEm.thunk.0
   >>> defined in <internal>
   >>> defined in <internal>

   ld64.lld: error: duplicate symbol: __ZN15partition_alloc8internal12_GLOBAL__N_126FreelistCorruptionDetectedEm.thunk.0
   >>> defined in <internal>
   >>> defined in <internal>

   ld64.lld: error: duplicate symbol: __ZN15partition_alloc8internal12_GLOBAL__N_126FreelistCorruptionDetectedEm.thunk.0
   >>> defined in <internal>
   >>> defined in <internal>
   ```

The input is probably invalid in some way, but the error is very cryptic and not helpful.

**Comment by nico** [comment]
```
 % git diff
diff --git a/lld/MachO/ConcatOutputSection.cpp b/lld/MachO/ConcatOutputSection.cpp
index 4cec418e7cf1..360452b00c85 100644
--- a/lld/MachO/ConcatOutputSection.cpp
+++ b/lld/MachO/ConcatOutputSection.cpp
@@ -324,7 +324,7 @@ void ConcatOutputSection::finalize() {
       StringRef thunkName = saver().save(funcSym->getName() + ".thunk." +
                                          std::to_string(thunkInfo.sequence++));
       r.referent = thunkInfo.sym = symtab->addDefined(
-          thunkName, /*file=*/nullptr, thunkInfo.isec, /*value=*/0,
+          thunkName, /*file=*/funcSym->getFile(), thunkInfo.isec, /*value=*/0,
           /*size=*/thunkSize, /*isWeakDef=*/false, /*isPrivateExtern=*/true,
           /*isThumb=*/false, /*isReferencedDynamically=*/false,
           /*noDeadStrip=*/false, /*isWeakDefCanBeHidden=*/false);
```

results in

```
% ~/src/llvm-project/out/gn/bin/ld64.lld @response.txt
ld64.lld: error: duplicate symbol: __ZN15partition_alloc8internal12_GLOBAL__N_126FreelistCorruptionDetectedEm.thunk.0
>>> defined in Users/thakis/src/chrome/src/out/gn/obj/third_party/blink/renderer/core/libblink_core.a(array_buffer_contents.o)
>>> defined in Users/thakis/src/chrome/src/out/gn/obj/third_party/blink/renderer/platform/libblink_platform.a(parkable_string.o)

ld64.lld: error: duplicate symbol: __ZN15partition_alloc8internal12_GLOBAL__N_126FreelistCorruptionDetectedEm.thunk.0
>>> defined in Users/thakis/src/chrome/src/out/gn/obj/third_party/blink/renderer/platform/libblink_platform.a(parkable_string.o)
>>> defined in Users/thakis/src/chrome/src/out/gn/obj/third_party/blink/renderer/platform/libblink_platform.a(skia_utils.o)

ld64.lld: error: duplicate symbol: __ZN15partition_alloc8internal12_GLOBAL__N_126FreelistCorruptionDetectedEm.thunk.0
>>> defined in Users/thakis/src/chrome/src/out/gn/obj/third_party/blink/renderer/platform/libblink_platform.a(skia_utils.o)
>>> defined in Users/thakis/src/chrome/src/out/gn/obj/third_party/blink/renderer/modules/mediasource/libmediasource.a(source_buffer.o)
thakis@Nicos-MacBook-Pro repro % demumble ' __ZN15partition_alloc8internal12_GLOBAL__N_126FreelistCorruptionDetectedEm'
 __ZN15partition_alloc8internal12_GLOBAL__N_126FreelistCorruptionDetectedEm
thakis@Nicos-MacBook-Pro repro % demumble '__ZN15partition_alloc8internal12_GLOBAL__N_126FreelistCorruptionDetectedEm' 
partition_alloc::internal::(anonymous namespace)::FreelistCorruptionDetected(unsigned long)
```

That's probably better? Still fairly cryptic though.

**Comment by nico** [comment]
The problem (upstream: crbug.com/1309891) is that a .h file has something like:

```
namespace {

[[noreturn]] NOINLINE void FreelistCorruptionDetected(size_t extra) {
  // Make it visible in minidumps.
  PA_DEBUG_DATA_ON_STACK("extra", extra);
  IMMEDIATE_CRASH();
}

}  // namespace
```

in a **`.h`** file. That's a TU-local function, so its thunk must also be TU-local. But we currently always put thunks into the symtab as private_extern. I think that's incorrect in this case.

**Comment by nico** [comment]
This makes things link:

```
% git diff
diff --git a/lld/MachO/ConcatOutputSection.cpp b/lld/MachO/ConcatOutputSection.cpp
index 4cec418e7cf1..a677419498a3 100644
--- a/lld/MachO/ConcatOutputSection.cpp
+++ b/lld/MachO/ConcatOutputSection.cpp
@@ -323,11 +323,20 @@ void ConcatOutputSection::finalize() {
 
       StringRef thunkName = saver().save(funcSym->getName() + ".thunk." +
                                          std::to_string(thunkInfo.sequence++));
-      r.referent = thunkInfo.sym = symtab->addDefined(
-          thunkName, /*file=*/nullptr, thunkInfo.isec, /*value=*/0,
-          /*size=*/thunkSize, /*isWeakDef=*/false, /*isPrivateExtern=*/true,
-          /*isThumb=*/false, /*isReferencedDynamically=*/false,
-          /*noDeadStrip=*/false, /*isWeakDefCanBeHidden=*/false);
+      if (!isa<Defined>(funcSym) || cast<Defined>(funcSym)->isExternal()) {
+        r.referent = thunkInfo.sym = symtab->addDefined(
+            thunkName, /*file=*/nullptr, thunkInfo.isec, /*value=*/0,
+            thunkSize, /*isWeakDef=*/false, /*isPrivateExtern=*/true,
+            /*isThumb=*/false, /*isReferencedDynamically=*/false,
+            /*noDeadStrip=*/false, /*isWeakDefCanBeHidden=*/false);
+      } else {
+        r.referent = thunkInfo.sym = make<Defined>(
+            thunkName, /*file=*/nullptr, thunkInfo.isec, /*value=*/0,
+            thunkSize, /*isWeakDef=*/false, /*isExternal=*/false,
+            /*isPrivateExtern=*/true, /*isThumb=*/false,
+            /*isReferencedDynamically=*/false, /*noDeadStrip=*/false,
+            /*isWeakDefCanBeHidden=*/false);
+      }
       thunkInfo.sym->used = true;
       target->populateThunk(thunkInfo.isec, funcSym);
       finalizeOne(thunkInfo.isec);
```

Need to add a test, then I'll send that out.

**Comment by nico** [comment]
Fix: https://reviews.llvm.org/D122624

**Comment by nico** [comment]
I was wondering how we could possibly hit this: There would have to be 128 MiB between the static function and its caller. Since it's a static function, that would require a .o file larger than 128 MiB, which seems unlikely!

Turns out the static function in one TU was ICF'd with a static function in another TU far away. The problem does not repro with `--icf=none`. And it _is_ believable that a byte-identical local function in another TU is over 128 MiB away.

**Comment by nico** [comment]
Fixed in 10cda6e36c05f83c8893e953b7a0f39ca8994ea3
